packages/dds/task-manager/src/taskManager.ts

```diff
@@ -0,0 +1,413 @@
+ * });
+ * ```
+ *
+ * These can be useful if the logic to volunteer for a task is separated from the logic to perform the task and it's
+ * not convenient to pass the Promise around.
+ */
+export class TaskManager extends SharedObject<ITaskManagerEvents> implements ITaskManager {
+    /**
+     * Create a new TaskManager
+     *
+     * @param runtime - data store runtime the new task queue belongs to
+     * @param id - optional name of the task queue
+     * @returns newly create task queue (but not attached yet)
+     */
+    public static create(runtime: IFluidDataStoreRuntime, id?: string) {
+        return runtime.createChannel(id, TaskManagerFactory.Type) as TaskManager;
+    }
+
+    /**
+     * Get a factory for TaskManager to register with the data store.
+     *
+     * @returns a factory that creates and load TaskManager
+     */
+    public static getFactory(): IChannelFactory {
+        return new TaskManagerFactory();
+    }
+
+    /**
+     * Mapping of taskId to a queue of clientIds that are waiting on the task.
+     */
+    private readonly taskQueues: Map<string, string[]> = new Map();
+
+    /**
+     * taskIds for tasks that we've sent a volunteer for but have not yet been ack'd.
+     */
+    private readonly pendingTaskQueues: Set<string> = new Set();
+
+    private readonly opWatcher: EventEmitter = new EventEmitter();
+    private readonly queueWatcher: EventEmitter = new EventEmitter();
+
+    /**
+     * Constructs a new task manager. If the object is non-local an id and service interfaces will
+     * be provided
+     *
+     * @param runtime - data store runtime the task queue belongs to
+     * @param id - optional name of the task queue
+     */
+    constructor(id: string, runtime: IFluidDataStoreRuntime, attributes: IChannelAttributes) {
+        super(id, runtime, attributes);
+
+        this.opWatcher.on("volunteer", (taskId: string, clientId: string) => {
+            this.addClientToQueue(taskId, clientId);
+        });
+
+        this.opWatcher.on("abandon", (taskId: string, clientId: string) => {
+            this.removeClientFromQueue(taskId, clientId);
+        });
+
+        runtime.getQuorum().on("removeMember", (clientId: string) => {
+            this.removeClientFromAllQueues(clientId);
+        });
+
+        this.queueWatcher.on("queueChange", (taskId: string, oldLockHolder: string, newLockHolder: string) => {
+            if (this.runtime.clientId === undefined) {
+                // TODO handle disconnected case
+                return;
+            }
+
+            if (oldLockHolder !== this.runtime.clientId && newLockHolder === this.runtime.clientId) {
+                this.emit("assigned", taskId);
+            } else if (oldLockHolder === this.runtime.clientId && newLockHolder !== this.runtime.clientId) {
+                this.emit("lost", taskId);
+            }
+        });
+    }
+
+    // TODO Remove or hide from interface, this is just for debugging
+    public getTaskQueues() {
```

Comment:
nit: consider prefixing 'private' things with an underscore so if we forget the TODO, consumers of the API have some indication that it's not meant for use.

Reply:
Changed in latest.